### PR TITLE
Add API to query collaborators permission for a repository

### PR DESCRIPTION
**Pull request by flozzone**
Targeting #14936, #15332

Adds a collaborator permissions API endpoint according to GitHub API: https://docs.github.com/en/rest/collaborators/collaborators#get-repository-permissions-for-a-user to retrieve a collaborators permissions for a specific repository.

### Checks the repository permissions of a collaborator. 

`GET` `/repos/{owner}/{repo}/collaborators/{collaborator}/permission`

Possible `permission` values are `admin`, `write`, `read`, `owner`, `none`.

```json
{
  "permission": "admin",
  "role_name": "admin",
  "user": {}
}
```

Where `permission` and `role_name` hold the same `permission` value and `user` is filled with the user API object. Only admins are allowed to use this API endpoint.

### CODE REVIEW

**Comment by flozzone** [comment]
GitHub API endpoint returns the permissions `admin`, `write`, `read`, and `none`. But Gitea also has `owner` in `AccessMode`. Should we just map the `owner` permission to the `admin` permission to be compatible with GitHub API or should we return the `owner` permission if the given user is also a owner of the repository?

Since this is a design question, I would really be happy to get a input from a author.

**Comment by lunny** [comment]
> GitHub API endpoint returns the permissions `admin`, `write`, `read`, and `none`. But Gitea also has `owner` in `AccessMode`. Should we just map the `owner` permission to the `admin` permission to be compatible with GitHub API or should we return the `owner` permission if the given user is also a owner of the repository?
> 
> Since this is a design question, I would really be happy to get a input from a author.

I vote to have an owner permission which should be equal to admin in Github.

**Comment by flozzone** [comment]
The current solution is to return the serialized form of `AccessMode`, which returns either: `admin`, `write`, `read`, `owner`, `none`.

I'll mention this in the PR description.

File: routers/api/v1/api.go [hunk 1]
```diff
@@ -801,9 +801,12 @@ func Routes() *web.Route {
...
 				m.Group("/collaborators", func() {
 					m.Get("", reqAnyRepoReader(), repo.ListCollaborators)
-					m.Combo("/{collaborator}").Get(reqAnyRepoReader(), repo.IsCollaborator).
-						Put(reqAdmin(), bind(api.AddCollaboratorOption{}), repo.AddCollaborator).
-						Delete(reqAdmin(), repo.DeleteCollaborator)
+					m.Group("/{collaborator}", func() {
+						m.Combo("").Get(reqAnyRepoReader(), repo.IsCollaborator).
+							Put(reqAdmin(), bind(api.AddCollaboratorOption{}), repo.AddCollaborator).
+							Delete(reqAdmin(), repo.DeleteCollaborator)
+						m.Get("/permission", repo.GetRepoPermissions)
```
Comment (lunny) [inline routers/api/v1/api.go:817]: permission check is necessary.

**Review by lunny — DISMISSED** [review]
see my comment

File: routers/api/v1/api.go (hunk 1 above)
Comment (flozzone) [reply routers/api/v1/api.go:817]: Added `reqAdmin()` to the `/permission` endpoint and added a hint to the PRs description.

File: routers/api/v1/api.go [hunk 2]
```diff
@@ -801,9 +801,12 @@ func Routes() *web.Route {
...
 				m.Group("/collaborators", func() {
 					m.Get("", reqAnyRepoReader(), repo.ListCollaborators)
-					m.Combo("/{collaborator}").Get(reqAnyRepoReader(), repo.IsCollaborator).
-						Put(reqAdmin(), bind(api.AddCollaboratorOption{}), repo.AddCollaborator).
-						Delete(reqAdmin(), repo.DeleteCollaborator)
+					m.Group("/{collaborator}", func() {
+						m.Combo("").Get(reqAnyRepoReader(), repo.IsCollaborator).
+							Put(reqAdmin(), bind(api.AddCollaboratorOption{}), repo.AddCollaborator).
+							Delete(reqAdmin(), repo.DeleteCollaborator)
+						m.Get("/permission", reqAdmin(), repo.GetRepoPermissions)
```
Comment (6543) [inline routers/api/v1/api.go]: i would check permission in repo.GetRepoPermissions,

so we can check for: `ctx.Doer.IsAdmin || ctx.Doer.ID == collaborator.ID`

File: integrations/api_repo_collaborator_test.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
```
Comment (6543) [inline integrations/api_repo_collaborator_test.go]:
```suggestion
// Copyright 2022 The Gitea Authors. All rights reserved.
```

File: models/repo_collaboration.go [hunk 4]
```diff
@@ -279,3 +279,18 @@ func IsOwnerMemberCollaborator(repo *repo_model.Repository, userID int64) (bool,
 
 	return db.GetEngine(db.DefaultContext).Get(&Collaboration{RepoID: repo.ID, UserID: userID})
 }
+
+// GetRepoCollaboratorAccessMode gets the repository access mode for the given user
+func GetRepoCollaboratorAccessMode(repo *repo_model.Repository, userID int64) (perm.AccessMode, error) {
```
Comment (6543) [inline models/repo_collaboration.go]: we dont have a func already for this?!?

File: routers/api/v1/api.go (hunk 1 above)
Comment (6543) [reply routers/api/v1/api.go:817]: https://github.com/go-gitea/gitea/pull/18761#discussion_r859726033

File: routers/api/v1/api.go (hunk 2 above)
Comment (lunny) [reply routers/api/v1/api.go]: And the sequence is wrong.

Comment (flozzone) [reply routers/api/v1/api.go]: > i would check permission in repo.GetRepoPermissions,
> 
> so we can check for: `ctx.Doer.IsAdmin || ctx.Doer.ID == collaborator.ID`

I will try to find a existing example

> And the sequence is wrong.

sequence of arguments for `m.Get`?

File: routers/api/v1/repo/collaborators.go [hunk 5]
```diff
@@ -233,6 +233,60 @@ func DeleteCollaborator(ctx *context.APIContext) {
...
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/RepoCollaboratorPermission"
+	//   "404":
+	//     "$ref": "#/responses/notFound"
+
+	// only admin and user for itself can change subscription
+	if ctx.Doer.IsAdmin || ctx.Doer.LoginName == ctx.Params(":collaborator") {
```
Comment (6543) [inline routers/api/v1/repo/collaborators.go]: sorry for the late suggsetstion, but doer with repo admin access should also be able to, in my opinion ...

if you need help, just tell.

File: models/repo_collaboration.go (hunk 4 above)
Comment (6543) [reply models/repo_collaboration.go]: we can delete this func and use p, e := GetUserRepoPermission(repo, user) directly

Comment (flozzone) [reply models/repo_collaboration.go]: can't really find one, any suggestion? `GetUserRepoPermission` is used in many places ;)

File: routers/api/v1/repo/collaborators.go [hunk 6]
```diff
@@ -233,6 +233,60 @@ func DeleteCollaborator(ctx *context.APIContext) {
...
+	if err != nil {
+		if user_model.IsErrUserNotExist(err) {
+			ctx.Error(http.StatusNotFound, "GetUserByName", err)
+		} else {
+			ctx.Error(http.StatusInternalServerError, "GetUserByName", err)
+		}
+		return
+	}
+
+	accessMode, err := models.GetRepoCollaboratorAccessMode(ctx.Repo.Repository, collaborator.ID)
```
Comment (6543) [inline routers/api/v1/repo/collaborators.go]:
```suggestion
	accessMode, err := models.GetUserRepoPermission(ctx.Repo.Repository, collaborator)
```

**Comment by 6543** [comment]
```sh
git checkout -f master -- models/repo_collaboration.go
git commit -S -m 'rm GetRepoCollaboratorAccessMode'
git push
```

File: models/repo_collaboration.go (hunk 4 above)
Comment (flozzone) [reply models/repo_collaboration.go]: I must have been blind

File: routers/api/v1/repo/collaborators.go [hunk 7]
```diff
@@ -233,6 +233,60 @@ func DeleteCollaborator(ctx *context.APIContext) {
...
+	//   required: true
+	// - name: collaborator
+	//   in: path
+	//   description: username of the collaborator
+	//   type: string
+	//   required: true
+	// responses:
+	//   "200":
+	//     "$ref": "#/responses/RepoCollaboratorPermission"
+	//   "404":
```
Comment (6543) [inline routers/api/v1/repo/collaborators.go:262]: response http.StatusForbidden not documented

File: models/repo_collaboration.go [hunk 8]
```diff
@@ -260,6 +259,20 @@ func reconsiderWatches(ctx context.Context, repo *repo_model.Repository, uid int
 	return removeIssueWatchersByRepoID(db.GetEngine(ctx), uid, repo.ID)
 }
 
+func getRepoTeams(e db.Engine, repo *repo_model.Repository) (teams []*Team, err error) {
```
Comment (6543) [inline models/repo_collaboration.go]: unrelated ?!?!

Comment (6543) [reply models/repo_collaboration.go]: I guess, you have an outdated main aka. master

Comment (6543) [reply models/repo_collaboration.go]: either "allow maintainer to edit" or add upstream remote and fetch changes :)

Comment (flozzone) [reply models/repo_collaboration.go]: its on top of `main` branch

Comment (6543) [reply models/repo_collaboration.go]: @flozzone `This branch is 2221 commits behind mgjm/gitea:master. ` ...

Comment (flozzone) [reply models/repo_collaboration.go]: edits by maintainers are allowed

Comment (6543) [reply models/repo_collaboration.go]: did you forked from another fork?

Comment (6543) [reply models/repo_collaboration.go]: tldr: you dont have to worry, just add upstream remote and use that

Comment (flozzone) [reply models/repo_collaboration.go]: why mgjm, need to counter check now

Comment (flozzone) [reply models/repo_collaboration.go]: i will squash my commits

Comment (6543) [reply models/repo_collaboration.go]:
```sh
git add remote upstream https://github.com/go-gitea/gitea
git fetch upstream
git checkout -f upstream/main -- models/repo_collaboration.go
...
```

Comment (6543) [reply models/repo_collaboration.go]: @flozzone you should not squash, that will only make reviews harder - it will be squash merged by us

Comment (flozzone) [reply models/repo_collaboration.go]: I was using your checkout CLI line, whose upstream didn't match with mine. what a wonder :)

File: integrations/api_repo_collaborator_test.go (hunk 3 above)
Comment (flozzone) [reply integrations/api_repo_collaborator_test.go]: Applied suggestion

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
`make generate-swagger`

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
> you should not squash, that will only make reviews harder - it will be squash merged by us

**Comment by flozzone** [comment]
thanks for holding my hand @6543 :people_holding_hands:

**Comment by flozzone** [comment]
> 


> you should not squash, that will only make reviews harder - it will be squash merged by us

Ok, I was actually not sure about this, I won't squash next time :+1:

File: integrations/api_repo_collaborator_test.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+	testCtx := NewAPITestContext(t, repo1Owner.Name, repo1.Name)
+
+	t.Run("RepoOwnerShouldBeAdmin", func(t *testing.T) {
+		req := NewRequestf(t, "GET", "/api/v1/repos/%s/%s/collaborators/%s/permission?token=%s", repo1Owner.Name, repo1.Name, repo1Owner.Name, testCtx.Token)
+		resp := session.MakeRequest(t, req, http.StatusOK)
+
+		var repoPermission api.RepoCollaboratorPermission
+		DecodeJSON(t, resp, &repoPermission)
+
+		assert.Equal(t, "owner", repoPermission.Permission)
```
Comment (6543) [inline integrations/api_repo_collaborator_test.go]:
```suggestion
		assert.Equal(t, "owner", repoPermission.Permission)
		assert.Equal(t,  ... repoPermission.RoleName)
```
...
:point_down: and below

**Comment by 6543** [comment]
you can test your tests localy too:

`make clean backend test-sqlite\#TestAPIRepoCollaboratorPermission`

File: integrations/api_repo_collaborator_test.go (hunk 9 above)
Comment (flozzone) [reply integrations/api_repo_collaborator_test.go]: well it now comes to the uncertainties that I had at the beginning of this PR.

Please see WhyHasUser5ReadAccess and WhyHasUser20ReadAccess tests

Comment (flozzone) [reply integrations/api_repo_collaborator_test.go]: It seems test-data is prefilled (or getting) with permissions for specific users, but I don't know which users

Comment (6543) [reply integrations/api_repo_collaborator_test.go]: test data is comming from https://github.com/go-gitea/gitea/tree/main/models/fixtures

in edgecases it could be the case, that the fixtures need a fix, as information that on real instances exist is missing ...

I'll have a look into it - just allow maintainers to edit the pull and i can push changes to this pull

Comment (flozzone) [reply integrations/api_repo_collaborator_test.go]: ok, thanks!

> I'll have a look into it - just allow maintainers to edit the pull and i can push changes to this pull

it is already allowed, at least that checkbox `Allow edits by maintainers` is checked.

Comment (6543) [reply integrations/api_repo_collaborator_test.go]: it could be an issue because it's a fork from an fork ... so `mgjm` is allowed but not me ... - github issues :/

Comment (6543) [reply integrations/api_repo_collaborator_test.go]: I'll file a pull against your repo if I found the issue

Comment (flozzone) [reply integrations/api_repo_collaborator_test.go]: oh I see, my bad, I can't change that now, without closing this PR.

> I'll file a pull against your repo if I found the issue

alright

btw. I don't have any experience in writing tests in go and I might use the test suite in a wrong way. `onGiteaRun` ?

**Review by 6543 — APPROVED** [review]
how could i overlooke that :sweat_smile:  - thats why tests are nice - works localy

File: routers/api/v1/repo/collaborators.go [hunk 10]
```diff
@@ -233,6 +233,62 @@ func DeleteCollaborator(ctx *context.APIContext) {
...
+	//     "$ref": "#/responses/notFound"
+	//   "403":
+	//     "$ref": "#/responses/forbidden"
+
+	// only admins can query all permissions, repo admins can query all repo permissions, collaborators can query only their own
+	if !ctx.Doer.IsAdmin && ctx.Doer.LoginName != ctx.Params(":collaborator") && !ctx.IsUserRepoAdmin() {
```
Comment (flozzone) [inline routers/api/v1/repo/collaborators.go:267]: the `LoginName` is missing for all users in the test fixtures , should I use `Name` instead? Or update test fixtures and use `LoginName`?

Comment (flozzone) [reply routers/api/v1/repo/collaborators.go:267]: I extended the fixtures

File: routers/api/v1/repo/collaborators.go [hunk 11]
```diff
@@ -233,6 +233,62 @@ func DeleteCollaborator(ctx *context.APIContext) {
...
+		ctx.Error(http.StatusForbidden, "User", "Only admins can query all permissions, repo admins can query all repo permissions, collaborators can query only their own")
+		return
+	}
+
+	collaborator, err := user_model.GetUserByName(ctx.Params(":collaborator"))
+	if err != nil {
+		if user_model.IsErrUserNotExist(err) {
+			ctx.Error(http.StatusNotFound, "GetUserByName", err)
```
Comment (flozzone) [inline routers/api/v1/repo/collaborators.go:275]: And this might be a security issue, since every repo admin would be able to check for user existence through 404 responses, even if the user is private! So we would leak sensible information.

Is this something I should consider?

Comment (6543) [reply routers/api/v1/repo/collaborators.go:275]: that's something we should add into GetUserByName - as it can be the case in many places !

I would say leave it as is and create a new issue specific to this!!!

Comment (flozzone) [reply routers/api/v1/repo/collaborators.go:275]: I wrote a mail to the security address as proposed in the docs

**Review by lunny — APPROVED** [review]

File: routers/api/v1/repo/collaborators.go [hunk 12]
```diff
@@ -233,6 +233,61 @@ func DeleteCollaborator(ctx *context.APIContext) {
...
+		} else {
+			ctx.Error(http.StatusInternalServerError, "GetUserByName", err)
+		}
+		return
+	}
+
+	permission, err := models.GetUserRepoPermission(ctx.Repo.Repository, collaborator)
```
Comment (6543) [inline routers/api/v1/repo/collaborators.go]:
```suggestion
	permission, err := models.GetUserRepoPermission(ctx, ctx.Repo.Repository, collaborator)
```

**Comment by 6543** [comment]
:rocket:
